hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java

```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
   }
 
   @Override
-  protected void processNextDeletedKey(HoodieKey hoodieKey) {
-    records.put(hoodieKey.getRecordKey(), SpillableMapUtils.generateEmptyPayload(hoodieKey.getRecordKey(),
-        hoodieKey.getPartitionPath(), getPayloadClassFQN()));
+  protected void processNextDeletedRecord(DeleteRecord deleteRecord) {
+    String key = deleteRecord.getRecordKey();
+    if (records.containsKey(key)) {
+      // Merge and store the merged record. The ordering val is taken to decide whether the same key record
+      // should be deleted or be kept. The old record is kept only if the DELETE record has smaller ordering val.
+      // For same ordering values, uses the natural order(arrival time semantics).
+
+      HoodieRecord<? extends HoodieRecordPayload> oldRecord = records.get(key);
+      Comparable curOrderingVal = oldRecord.getData().getCombineValue();
+      Comparable deleteOrderingVal = deleteRecord.getOrderingVal();
+      // Checks the ordering value does not equal to 0
+      // because we use 0 as the default value which means natural order
+      boolean choosePrev = !deleteOrderingVal.equals(0)
+          && curOrderingVal.getClass() == deleteOrderingVal.getClass()
```

Comment:
curOrderingVal.getClass().getName().equals(deleteOrderingVal.getClass().getName())

Reply:
![image](https://user-images.githubusercontent.com/36881649/159112163-bf0e9804-e0a5-4665-9c8d-1a9783c2ff64.png)

Reply:
+1. we can add a utiltiy to ReflectUtils or ClassUtils for the comparison

Reply:
can you help me understand how this would pan out for records written with 0.10.0 or before. 
I would assume DeleteRecord will have 0 as ordering value, where as, an insert record will have proper ordering value, lets say 10. 

so, if we get an insert followed by delete, we will choose delete record as final value. 
and if we get a delete followed by an insert, we will choose insert record as the final value.

Reply:
Before the patch, both `DELETE` -> `INSERT`  and `INSERT` -> `DELETE` choose `DELETE`, with this patch, `DELETE` is only chosen if it has a bigger ordering value.

Reply:
For same JVM classloader, use `==` is more efficient BTW, i can move it to `ReflectionUtils`.